enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,325 @@
+patterns.
+
+The following IngressController enables mandatory use of client certificates,
+verified using the CA certificates in the `router-ca-certs-default` configmap,
+and requiring that client certificates match the provided subject:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  clientTLS:
+    clientCertificatePolicy: Required
+    clientCA:
+      name: router-ca-certs-default
+    allowedSubjectPatterns:
+    - "^/CN=example.com/ST=NC/C=US/O=Security/OU=OpenShift$"
+```
+
+#### Validation
+
+Omitting `spec.clientTLS` has well defined semantics.
+
+The API validates the `spec.clientTLS.clientCertificatePolicy` field value as
+described by the field type's `+kubebuilder:validation:Enum` marker.
+
+The API does not validate that the `spec.clientTLS.clientCA.name` field value
+references a valid configmap; if the named configmap does not exist, then the
+IngressController deployment is not scheduled until the configmap is created.
+
+The API does not validate the `spec.clientTLS.allowedSubjectPatterns` field
+value.
+
+### User Stories
+
+#### As a cluster administrator, I do not want to enable use of client certificates.
+
+Client TLS is not enabled by default, so no action is needed.
+
+#### As a cluster administrator, I want to enable optional use of client certificates.
+
+The user enables client TLS with the `Optional` client certificate policy.  The
+operator configures the IngressController to request and verify client
+certificates using the system trust bundle.
+
+#### As a cluster administrator, I want to require client certificates using a CA that I control.
+
+The user enables client TLS with the `Required` client certificate policy and a
+custom CA certificate bundle.  The operator configures the IngressController to
+request and verify client certificates using the provided CA certificate bundle
+and reject clients that do not present valid client certificates.
+
+#### As a cluster administrator, I want to require client certificates using a CA that I control, respecting its certificate revocation list.
+
+The user enables client TLS with the `Required` client certificate policy and a
+custom CA certificate bundle with a CA certificate using the X509v3 extension to
+specify one or more CRL destination points in the certificate.  The operator
+configures the IngressController to request and verify client certificates using
+the provided CA certificate bundle and reject clients that do not present valid
+client certificates.  The operator periodically updates the CRL using the
+destination points.
+
+#### As a cluster administrator, I want to force a refresh of all certificate revocation lists.
+
+The user deletes the `router-client-ca-crl-<name>` configmap in the
+`openshift-ingress` namespace.  The deletion triggers the operator's
+reconciliation logic, which redownloads the CRLs and recreates the configmap.
```

Comment:
What happens if the CRL is too big to be stored in a configMap? They are often well over 1MB in size.
Providing the option to read a local (storage volume?) CRL thus ignoring CRL distribution points in the CA would be useful for customers that have to lug a huge CRL around.

Reply:
Right, we need to document storage limitations, and allowing use of a PVC makes sense.  Something like the following?  
```go
const (
	CertificateRevocationListAutoSourceType CertificateRevocationListSourceType = ""

	CertificateRevocationListConfigMapSourceType CertificateRevocationListSourceType = "ConfigMap"

	CertificateRevocationListPersistentVolumeSourceType CertificateRevocationListSourceType = "PersistentVolume"
)

// +union
type CertificateRevocationList struct {
	// +unionDiscriminator
	// +kubebuilder:validation:Required
	// +required
	Source CertificateRevocationListSourceType `json:"source"`

	// persistentVolumeClaim is the name of the Persisent Volume Claim for a
	// volume that contains the CRL.  The CRL must be stored in PEM format
	// in a file named crl.pem in the volume's root directory.
	//
	// +optional
	PersistentVolumeClaim string `json:"persistentVolumeClaim,omitempty"`

	// configMap is a reference to a ConfigMap to be used.
	//
	// +optional
	ConfigMap configv1.ConfigMapNameReference `json:"configMap"`
}

type ClientTLS struct {
	// ...

	// certificateRevocationList specifies how to get the certificate
	// revocation list (CRL) for a client CA certificate.  If this field is
	// empty, each certificate in the client CA certificate bundle is
	// checked for distribution points, which are then used to download the
	// CRL.  Any and all downloaded CRLs must be able to fit within a
	// ConfigMap.
        //
	// +required
        CertificateRevocationList json:"certificateRevocationList"

	// ...
}
```

Reply:
I would avoid using a PVC.  You're not talking about that much data.  What format is the CRL in?  Is it splittable?

Reply:
I was planning to use `x509.ParseCRL`, which can handle PEM- or DER-encoded CRLs.  I don't think it's splittable (which is ironic given the existence of OCSP stapling!).